tools/xmlschema.py

```diff
@@ -0,0 +1,462 @@
+from xml.etree import ElementTree
+from copy import deepcopy
+from itertools import permutations, zip_longest
+
+from pathlib import Path
+import argparse
+from typing import List
+
+
+cmd_arg_parser = argparse.ArgumentParser(
+    description="Create an XML schema from a SDFormat file."
+)
+cmd_arg_parser.add_argument(
+    "-i",
+    "--in",
+    dest="source",
+    required=True,
+    type=str,
+    help="SDF file inside of directory to compile.",
+)
+cmd_arg_parser.add_argument(
+    "-s",
+    "--sdf",
+    dest="directory",
+    required=True,
+    type=str,
+    help="Directory containing all the SDF files.",
+)
+cmd_arg_parser.add_argument(
+    "-o",
+    "--out",
+    dest="target",
+    default=".",
+    type=str,
+    help="Output directory for xsd file. Will be created if it doesn't exit.",
+)
+cmd_arg_parser.add_argument(
+    "--ns-prefix",
+    dest="ns_prefix",
+    default="http://sdformat.org/schema",
+    type=str,
+    help="Prefix for generated xsd namespaces.",
+)
+
+args = cmd_arg_parser.parse_args()
+source_dir = Path(args.directory)
+source: Path = source_dir / args.source
+
+root = ElementTree.parse(source).getroot()
+declared = dict()
+declared["imports"] = set()
+
+
+def _to_simple_type(in_type: str) -> str:
+    known_types = {
+        "unsigned int": "xs:unsignedInt",
+        "unsigned long": "xs:unsignedLong",
+        "bool": "xs:boolean",
+        "string": "xs:string",
+        "double": "xs:double",
+        "int": "xs:int",
+        "float": "xs:float",
+        "char": "xs:char",
+        "vector3": "types:vector3",
+        "vector2d": "types:vector2d",
+        "vector2i": "types:vector2i",
+        "pose": "types:pose",
+        "time": "types:time",
+        "color": "types:color",
+    }
+
+    try:
+        return known_types[in_type]
+    except KeyError:
+        raise RuntimeError(f"Unknown simple type: {in_type}") from None
+
+
+# collect existing namespaces
+namespaces = {
+    "types": "http://sdformat.org/schemas/types.xsd",
+    "xs": "http://www.w3.org/2001/XMLSchema",
+}
+for path in source_dir.iterdir():
+    if not path.is_file():
+        continue
+
+    if not path.suffix == ".sdf":
+        continue
+
+    namespaces[path.stem] = f"{args.ns_prefix}/{path.stem}"
+    ElementTree.register_namespace(path.stem, f"{args.ns_prefix}/{path.stem}")
+
+
+def _to_qname(name: str) -> str:
+    try:
+        prefix, name = name.split(":")
+    except ValueError:
+        # no prefix
+        return name
+    else:
+        return f"{{{namespaces[prefix]}}}{name}"
+
+
+class Description:
+    def __init__(self, element: ElementTree.Element):
+        self.element = element
+
+    def to_subtree(self):
```

Comment:
Type annotations are indeed very nice; I wonder though if we want them here or if they will mainly add clutter. The reason you'd normally want them (in my mind) is for static validation via mypy or because you want to add IDE auto-completion/suggestions. Neither of these are what we are going for here (we just want a class that works in side the confines of this script).

@Bi0T1N  Do you think it is still worth having?

Reply:
I'm always for having type annotations because you never know who else is using your code and might benefit from it. Apart from that, there is also a security concern since Python is dynamically typed as you mentioned.